runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java

```diff
@@ -1224,10 +1225,10 @@ private static void translateFn(
       Map<TupleTag<?>, Coder<?>> outputCoders,
       DoFnSchemaInformation doFnSchemaInformation,
       Map<String, PCollectionView<?>> sideInputMapping) {
-    DoFnSignature signature = DoFnSignatures.getSignature(fn.getClass());
 
-    if (signature.usesState() || signature.usesTimers()) {
-      DataflowRunner.verifyStateSupported(fn);
+    boolean isStateful = DoFnFeatures.isStateful(fn);
```

Comment:
This is a good example of `DoFnFeatures` being useful: since "stateful ParDo" is really a totally separate primitive, it makes sense to just have a single boolean indicating it. Eventually moving so the different primitives are more clearly split with different URNs would be good.

(no action for you here - just a positive comment)